src/Lecture2.hs

```diff
@@ -287,5 +433,54 @@ x + 45 + y
 Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
+{- I think the simplest form possible here is (optionally) a constant, and a list of variables -}
+{- So I'm going to traverse the tree, collecting the literal value and vars to get the simplest form -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr = case simplifyExpr expr of
+   (Nothing, Nothing) -> Lit 0
+   (Just v1, Nothing) -> Lit v1
+   (Nothing, Just e2) -> e2
+   (Just v1, Just e2) -> Add (Lit v1) e2
+
+{- Auxiliary function - processes the tree recursively, bubbling up the constant and isolating the rest of expression-}
+simplifyExpr :: Expr -> (Maybe Int, Maybe Expr)
+simplifyExpr expr = case expr of
+   Lit 0     -> (Nothing, Nothing)
+   Lit value -> (Just value, Nothing)
+   Var v     -> (Nothing, Just expr)
+   Add expr1 expr2 -> let
+      res1 = simplifyExpr expr1
+      res2 = simplifyExpr expr2
+      joinedValues = joinValues (fst res1) (fst res2)
+      joinedExpressions = joinExpressions (snd res1) (snd res2)
+      in case joinedValues of
+        Nothing -> (Nothing, joinedExpressions)
+        Just 0  -> (Nothing, joinedExpressions)
+        Just v  -> (Just v, joinedExpressions)
+
+joinValues :: Maybe Int -> Maybe Int -> Maybe Int
+joinValues = join (\x y -> x+y)
```

Comment:
At first, I thought it was a standard function `join` for Monads and I spent quite a lot of time trying to figure out how it works 😅  And only after that I noticed that this is a helper function defined below 🙂